File: flutter_application/lib/core/services/meme_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';
import 'package:uuid/uuid.dart';
import 'package:path_provider/path_provider.dart';
import '../models/meme.dart';
import 'llm_service.dart';

class MemeService {
  static Database? _database;
  final LLMService _llmService = LLMService();
  final Uuid _uuid = const Uuid();

  Future<Database> get database async {
    if (_database != null) return _database!;
    _database = await _initDB('astra_me.db');
    return _database!;
  }

  Future<Database> _initDB(String filePath) async {
    final dbPath = await getDatabasesPath();
    final path = join(dbPath, filePath);

    return await openDatabase(path, version: 1, onOpen: (db) async {
      // Ensure table exists even if DB was created by MemoryService
      await db.execute('''
        CREATE TABLE IF NOT EXISTS memes (
          id TEXT PRIMARY KEY,
          path TEXT NOT NULL,
          tags TEXT NOT NULL,
          embedding TEXT NOT NULL,
          usage_count INTEGER DEFAULT 0,
          created_at INTEGER NOT NULL
        )
      ''');
    });
  }

  Future<void> saveMeme(String path, String description) async {
    List<double> embedding = [];
    try {
      embedding = await _llmService.getEmbedding(description);
    } catch (e) {
      print("Warning: Failed to generate embedding for meme: $e");
      // Fallback: use zero vector or skip? 
      // For now, we'll skip saving if we can't search it, or use a dummy vector.
      // Better to just return or throw.
      return;
    }

    try {
      final meme = Meme(
        id: _uuid.v4(),
        path: path,
        tags: description,
        embedding: embedding,
        createdAt: DateTime.now(),
      );

      final db = await database;
      await db.insert('memes', meme.toMap());
      print("Meme saved: $description");
    } catch (e) {
      print("Error saving meme to DB: $e");
    }
  }

  Future<void> saveMemeFromBytes(Uint8List bytes, String description) async {
    try {
      final appDir = await getApplicationDocumentsDirectory();
      final memeDir = Directory(join(appDir.path, 'memes'));
      if (!await memeDir.exists()) {
        await memeDir.create(recursive: true);
      }
      final filename = '${_uuid.v4()}.png';
      final file = File(join(memeDir.path, filename));
      await file.writeAsBytes(bytes);
      
      await saveMeme(file.path, description);
    } catch (e) {
      print("Error saving meme file: $e");
    }
  }

  Future<List<Meme>> searchMemes(String query, {int limit = 3}) async {
    try {
      final queryEmbedding = await _llmService.getEmbedding(query);
      final db = await database;
      final List<Map<String, dynamic>> maps = await db.query('memes');
      
      if (maps.isEmpty) return [];

      List<Meme> allMemories = maps.map((e) => Meme.fromMap(e)).toList();
      
      var scored = allMemories.map((meme) {
        double score = _cosineSimilarity(queryEmbedding, meme.embedding);
        return MapEntry(meme, score);
      }).toList();

      // Sort by score descending
      scored.sort((a, b) => b.value.compareTo(a.value));

      // Filter by threshold (e.g. 0.7) to avoid irrelevant memes
      // But for "fun", maybe lower threshold.
      return scored.where((e) => e.value > 0.6).take(limit).map((e) => e.key).toList();
    } catch (e) {
      print("Error searching memes: $e");
      return [];
    }
  }

  Future<void> incrementUsage(String id) async {
    final db = await database;
    await db.rawUpdate('UPDATE memes SET usage_count = usage_count + 1 WHERE id = ?', [id]);
  }

  double _cosineSimilarity(List<double> a, List<double> b) {
    if (a.length != b.length) return 0.0;
    double dotProduct = 0.0;
    double normA = 0.0;
    double normB = 0.0;

    for (int i = 0; i < a.length; i++) {
      dotProduct += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }

    if (normA == 0 || normB == 0) return 0.0;
    return dotProduct / (sqrt(normA) * sqrt(normB));
  }
}

File: backend/app/services/meme_service.py
import json
import math
import uuid
from typing import List
from datetime import datetime
from sqlmodel import Session, select
from app.models.database import engine
from app.models.meme import Meme
from app.services.llm_service import LLMService

class MemeService:
    def __init__(self):
        self.llm_service = LLMService()

    async def save_meme(self, path: str, description: str, api_key: str = None, base_url: str = None) -> Meme:
        """
        Save a new meme to the database with embedding.
        """
        embedding = await self.llm_service.get_embedding(description, api_key=api_key, base_url=base_url)
        
        meme = Meme(
            id=str(uuid.uuid4()),
            path=path,
            description=description,
            embedding=json.dumps(embedding),
            created_at=datetime.now(),
            usage_count=0
        )
        
        with Session(engine) as session:
            session.add(meme)
            session.commit()
            session.refresh(meme)
            
        return meme

    async def search_memes(self, query: str, limit: int = 1, threshold: float = 0.6, api_key: str = None, base_url: str = None) -> List[Meme]:
        """
        Search for relevant memes based on semantic similarity.
        """
        try:
            query_embedding = await self.llm_service.get_embedding(query, api_key=api_key, base_url=base_url)
            
            with Session(engine) as session:
                memes = session.exec(select(Meme)).all()
                
                scored_memes = []
                for meme in memes:
                    score = self._cosine_similarity(query_embedding, meme.embedding_list)
                    if score >= threshold:
                        scored_memes.append((meme, score))
                
                # Sort by score desc
                scored_memes.sort(key=lambda x: x[1], reverse=True)
                
                return [m[0] for m in scored_memes[:limit]]
        except Exception as e:
            print(f"[MemeService] Search error: {e}")
            return []

    def _cosine_similarity(self, vec_a: List[float], vec_b: List[float]) -> float:
        if not vec_a or not vec_b or len(vec_a) != len(vec_b):
            return 0.0
        
        dot_product = sum(a * b for a, b in zip(vec_a, vec_b))
        norm_a = math.sqrt(sum(a * a for a in vec_a))
        norm_b = math.sqrt(sum(b * b for b in vec_b))
        
        if norm_a == 0 or norm_b == 0:
            return 0.0
            
        return dot_product / (norm_a * norm_b)

    def increment_usage(self, meme_id: str):
        with Session(engine) as session:
            meme = session.get(Meme, meme_id)
            if meme:
                meme.usage_count += 1
                session.add(meme)
                session.commit()
